superset-frontend/src/middleware/asyncEvent.ts

```diff
@@ -0,0 +1,194 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { Middleware, MiddlewareAPI, Dispatch } from 'redux';
+import { makeApi, SupersetClient } from '@superset-ui/core';
+import { SupersetError } from 'src/components/ErrorMessage/types';
+import { getFeatureFlag, isFeatureEnabled, FeatureFlag } from '../featureFlags';
+import {
+  getClientErrorObject,
+  parseErrorJson,
+} from '../utils/getClientErrorObject';
+
+export type AsyncEvent = {
+  id: string;
+  channel_id: string;
+  job_id: string;
+  user_id: string;
+  status: string;
+  errors: SupersetError[];
+  result_url: string;
+};
+
+type AsyncEventOptions = {
+  getPendingComponents: (state: any) => any[];
+  successAction: (componentId: number, componentData: any) => { type: string };
+  errorAction: (componentId: number, response: any) => { type: string };
+  processEventsCallback?: (events: AsyncEvent[]) => void; // this is currently used only for tests
+};
+
+type CachedDataResponse = {
+  componentId: number;
+  status: string;
+  data: any;
+};
+
+const initAsyncEvents = (options: AsyncEventOptions) => {
+  // TODO: implement websocket support
+  const TRANSPORT_POLLING = 'polling';
+  const config = getFeatureFlag(FeatureFlag.GLOBAL_ASYNC_QUERIES_OPTIONS) || {};
+  const transport = config.transport || TRANSPORT_POLLING;
+  const polling_delay = config.polling_delay || 500;
+  const {
+    getPendingComponents,
+    successAction,
+    errorAction,
+    processEventsCallback,
+  } = options;
+
+  const middleware: Middleware = <S>(store: MiddlewareAPI<S>) => (
+    next: Dispatch<S>,
+  ) => {
+    const JOB_STATUS = {
+      PENDING: 'pending',
+      RUNNING: 'running',
+      ERROR: 'error',
+      DONE: 'done',
+    };
+    const LOCALSTORAGE_KEY = 'last_async_event_id';
+    const POLLING_URL = '/api/v1/async_event/';
+    let lastReceivedEventId: string | null;
+
+    try {
+      lastReceivedEventId = localStorage.getItem(LOCALSTORAGE_KEY);
+    } catch (err) {
+      console.warn('failed to fetch last event Id from localStorage');
+    }
+
+    const fetchEvents = makeApi<
+      { last_id?: string | null },
+      { result: AsyncEvent[] }
+    >({
+      method: 'GET',
+      endpoint: POLLING_URL,
+    });
+
+    const fetchCachedData = async (
+      asyncEvent: AsyncEvent,
+      componentId: number,
+    ): Promise<CachedDataResponse> => {
+      let status = 'success';
+      let data;
+      try {
+        const { json } = await SupersetClient.get({
+          endpoint: asyncEvent.result_url,
+        });
+        data = 'result' in json ? json.result[0] : json;
+      } catch (response) {
+        status = 'error';
+        data = await getClientErrorObject(response);
+      }
+
+      return { componentId, status, data };
+    };
+
+    const setLastId = (asyncEvent: AsyncEvent) => {
+      lastReceivedEventId = asyncEvent.id;
+      try {
+        localStorage.setItem(LOCALSTORAGE_KEY, lastReceivedEventId as string);
```

Comment:
Why use local storage here? Can't this just live in state? I foresee issues with users getting into odd states where a hard refresh of the browser won't clear things up.

Reply:
The async event cookie persists across page loads, while state does not. We don't want to fetch all (already processed) events in the stream from the start of the user's session, but only events from the last event received forward.

Reply:
Can this be stored in server side session data?

Reply:
In fact, I'm wondering whether the JWT can also be stored in server side session, too. Each auth user will have only one jwt/last received event id at a time, right?

Reply:
@robdiciuccio @dpgaspar do you know if we actively use [request.session](https://flask-session.readthedocs.io/en/latest/) for anything?